aircmd/models/base.py

```diff
@@ -45,50 +44,143 @@ class GlobalSettings(BaseSettings):
+    async def create(cls, name: str, parent_pipeline: Optional['Pipeline'] = None) -> 'Pipeline':
+        if parent_pipeline is None:
+            # This is a top-level pipeline. Get the Dagger client from the current PipelineContext.
+            async with PipelineContext() as ctx:
+                dagger_client = ctx.dagger_client
+        else:
+            # This is a nested pipeline. Get the Dagger client from the parent pipeline.
+            dagger_client = parent_pipeline.dagger_client
+
+        # Create the pipeline using the appropriate Dagger client.
+       
+        dagger_client = dagger_client.pipeline(name)
+        return cls(name=name, parent_pipeline=parent_pipeline, dagger_client=dagger_client)
+
+
 class PipelineContext(BaseModel, Singleton):
-    dagger_config: dagger.Config
-    dagger_connection: dagger.Connection
+    _current_level: int = 0
+    logger: logging.Logger
+    current_running_tasks: int = 0
+    max_concurrency: int = 0
+    max_concurrency_per_level: dict[int, int] = {}
+    concurrency_lock: anyio.Lock = anyio.Lock()
+    _dagger_client: Optional[dagger.Client] = PrivateAttr()
+    dagger_client: dagger.Client = Field(default=None, init=False)
 
-    def __init__(self, config: Optional[dagger.Config] = None,  
-                 connection: Optional[dagger.Connection] = None, 
-                 **data: Any):
-        if config is None:
-            config = dagger.Config(log_output=sys.stdout)
-        if connection is None:
-            # Replace with your logic to create a Connection instance
-            connection = dagger.Connection(config = config)
-        super().__init__(dagger_config=config, dagger_connection=connection, **data)
 
     class Config:
-        arbitrary_types_allowed = True
+        arbitrary_types_allowed=True
+
+    
+    async def __aenter__(self) -> 'PipelineContext':
+        self.dagger_client = await dagger.Connection(dagger.Config(log_output=sys.stdout)).__aenter__()
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
+        if self.dagger_client is not None:
+            pass
+
+
+    async def update_concurrency(self, delta: int, level: int) -> None:
+        async with self.concurrency_lock:
+            self.current_running_tasks += delta
+            current_level_concurrency = self.max_concurrency_per_level.get(level, 0) + delta
+            self.max_concurrency_per_level[level] = max(self.max_concurrency_per_level.get(level, 0), current_level_concurrency)
+            self.max_concurrency = max(self.max_concurrency_per_level.values())
+
+
+    async def run_pipelines(self, pipelines: List[Coroutine], concurrency: int) -> None:
+        semaphore = anyio.Semaphore(concurrency)
+
+        async def run_pipeline_with_semaphore(pipeline: Coroutine) -> None:
+            async with semaphore:
+                await self.update_concurrency(1, self._current_level)
+                await pipeline
```

Comment:
What happens if `pipeline` raises an exception and we don't decrement the concurrency? I suppose it doesn't matter since that would kill the whole pipeline, but perhaps we should handle this more gracefully e.g. wrap this in a `try/except`

Reply:
Agreed! We should still report accurate information at the end, and the way that I have it now precludes that

Also, I think that the behavior you mention is interesting about 'killing the whole pipeline' because @alafanechere and @bnchrch and I have found out that it's valuable to make that behavior very configurable for end users.

In other words it's interesting to support the following scenarios

- if one pipeline fails, halt execution of the entire tree branch
- if one pipeline fails, halt execution of the entire pipeline
- if one pipeline fails, halt execution of only the child pipelines of the failed pipeline

Reply:
This is what [cancel scopes](https://trio.readthedocs.io/en/latest/reference-core.html#cancellation-semantics) are for.